File: python/dynamic_programming/coin_change.py
"""
THE COIN CHANGE PROBLEM (SOLVED FROM HACKERRANK- PROBLEM SOLVING)
PROBLEM:
You are working at the cash counter at a fun-fair, and you have different types of coins available 
to you in infinite quantities. The value of each coin is already given.
 Can you determine the number of ways of making change for a particular number of units using the given types of coins?

For example, if you have 4 types of coins, and the value of each type is given as 8,3,1,2 respectively, 
you can make change for 3 units in three ways:{1,1,1} ,{1,2} and {3} .

Input Format

The first line contains two space-separated integers describing the respective values of n and m, where:
n is the number of units
m is the number of coin types
The second line contains m space-separated integers describing the respective values of each coin type :
c=c[0],c[1],c[2]...c[m-1]  (the list of distinct coins available in infinite amounts).

Output Format

Print a long integer denoting the number of ways we can get a sum of  n
from the given infinite supply of m types of coins

Sample Input 0
4 3
1 2 3
Sample Output 0
4
Explanation 0

There are four ways to make change for N=4 using coins with values given by C=[1,2,3] :
1. {1,1,1}
2. {1,1,2}
3. {2,2}
4. {1,3}

Thus, we print 4 as our answer.
"""

def getWays(n, c):
    n_perms = [1]+[0]*n
    for coin in c:
        for i in range(coin, n+1):
            n_perms[i] += n_perms[i-coin]
    return n_perms[n]
    
n, m = map(int, input().split())
c = list(map(int, input().split()))
print(getWays(n, c))

File: go/dynamic_programming/coin_change.go
/*
	Language: Go
	DP: Coin Change

	Given a value M, if we want to make change for M cents, and we have infinite supply of each of S = { S1, S2, .. , Sn} valued coins,
	how many ways can we make the change? The order of coins doesn’t matter.
	Example : For M = 10 and S = {2, 5, 3, 6}, there are five solutions: {2,2,2,2,2}, {2,2,3,3}, {2,2,6}, {2,3,5} and {5,5}. So the number of ways is 5.

*/

package main

import (
	"fmt"
)

var coin [1e5 + 7]int           // array for value of coins
var table [1e3 + 4][1e3 + 4]int // dp table, table[i][j] stores ways of making i cents using coins having values from first j values

func main() {
	var size, m int
	// reading array from user
	fmt.Println("Enter number of coins: ")
	fmt.Scan(&size)
	fmt.Println("Enter value to be paid (M): ")
	fmt.Scan(&m)
	fmt.Println("Enter value of coins sequenctially: ")
	for i := 0; i < size; i++ {
		fmt.Scan(&coin[i])
	}

	ans := CountWays(size, m)
	fmt.Printf("Number of possible ways to make %d cents = %d\n", m, ans)
}

// function to find number of possible ways to make 'm' cents
func CountWays(n int, m int) int {

	for i := 0; i < n; i++ {
		table[0][i] = 1 //there is 1 way to make 0 cents using first i coins
	}

	// bottom up filling
	for i := 1; i < m+1; i++ {
		for j := 0; j < n; j++ {
			// number of ways incuding atleast 1 coin[j]
			way1 := 0
			if i-coin[j] >= 0 {
				way1 = table[i-coin[j]][j]
			}
			// number of ways excluding coin[j]
			way2 := 0
			if j >= 1 {
				way2 = table[i][j-1]
			}
			table[i][j] = way1 + way2
		}
	}
	return table[m][n-1]
}

/*
	Sample Output:

	Enter number of coins:
	4
	Enter value to be paid (M):
	10
	Enter value of coins sequenctially:
	2 5 3 6
	Number of possible ways to make 10 cents = 5

*/
